### PR TITLE
F ros control

### LINKED ISSUE #32: Switch to ros_ethercat and ros_control

### PR DESCRIPTION
**Pull request by shadowmanos**
wait for the test with hardware before merging

merge at the same time as shadow-robot/sr-ros-interface#127

closed #32 

after merge clone shadow_robot/ros_ethercat

### CODE REVIEW

File: sr_edc_controller_configuration/scripts/controllers_to_diagnostics.py [hunk 1]
```diff
@@ -26,7 +26,7 @@
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 
-import roslib; roslib.load_manifest('pr2_controller_manager')
+import roslib
 import rospy
```
Comment (markpitchless) [inline sr_edc_controller_configuration/scripts/controllers_to_diagnostics.py:29]: You don't need the roslib import, rospy is enough under hydro catkin.

File: sr_edc_controller_configuration/sr_edc_controller_manager.launch [hunk 2]
```diff
@@ -12,8 +12,7 @@
 
 
 <!-- Controller manager parameters -->
-  <param name="pr2_controller_manager/mechanism_statistics_publish_rate" value="1.0" />
-  <param name="pr2_controller_manager/joint_state_publish_rate" value="100.0" />
+  <!-- param name="pr2_controller_manager/mechanism_statistics_publish_rate" value="1.0" /-->
```
Comment (markpitchless) [inline sr_edc_controller_configuration/sr_edc_controller_manager.launch]: What is happening here, are these options not supported by the new code? Is the commented out option a reminder or can it be removed?

File: sr_edc_ethercat_drivers/CMakeLists.txt [hunk 3]
```diff
@@ -4,7 +4,7 @@ cmake_minimum_required(VERSION 2.8.3)
 project(sr_edc_ethercat_drivers)
 # Load catkin and all dependencies required for this package
 # TODO: remove all from COMPONENTS that are not catkin packages.
-find_package(catkin REQUIRED COMPONENTS roscpp roslib std_msgs sr_robot_lib ethercat_hardware sr_robot_msgs sr_external_dependencies message_generation)
+find_package(catkin REQUIRED COMPONENTS roscpp roslib std_msgs sr_robot_lib ros_ethercat_model sr_robot_msgs sr_external_dependencies message_generation)
```
Comment (markpitchless) [inline sr_edc_ethercat_drivers/CMakeLists.txt]: Why the change of name from hardware to model?

**Comment by markpitchless** [comment]
This looks good to me, but I'd like at least a 2nd review before merge as this is core stuff.

File: sr_robot_lib/src/sr_muscle_robot_lib.cpp [hunk 4]
```diff
@@ -194,7 +192,7 @@ namespace shadow_robot
       for (; joint_tmp != this->joints_vector.end(); ++joint_tmp)
       {
         boost::shared_ptr<shadow_joints::MuscleWrapper> muscle_wrapper = boost::static_pointer_cast<shadow_joints::MuscleWrapper>(joint_tmp->actuator_wrapper);
-        sr_actuator::SrMuscleActuator* muscle_actuator = static_cast<sr_actuator::SrMuscleActuator*>(muscle_wrapper->actuator);
+        sr_actuator::SrMuscleActuator* muscle_actuator = dynamic_cast<sr_actuator::SrMuscleActuator*>(muscle_wrapper->actuator);
```
Comment (toliver) [inline sr_robot_lib/src/sr_muscle_robot_lib.cpp]: It is unnecessary to do a dynamic cast, as the object will always be of the right type.
Besides, we are not adding any checks to assess that the conversion was successful, so there is no real need for the change.

File: sr_robot_lib/src/sr_motor_robot_lib.cpp [hunk 5]
```diff
@@ -122,7 +120,7 @@ namespace shadow_robot
 
       this->process_position_sensor_data(joint_tmp, status_data, timestamp);
 
-      sr_actuator::SrActuator* motor_actuator = static_cast<sr_actuator::SrActuator*>(joint_tmp->actuator_wrapper->actuator);
+      sr_actuator::SrActuator* motor_actuator = dynamic_cast<sr_actuator::SrActuator*>(joint_tmp->actuator_wrapper->actuator);
```
Comment (toliver) [inline sr_robot_lib/src/sr_motor_robot_lib.cpp]: same here

File: sr_robot_lib/src/sr_motor_robot_lib.cpp [hunk 6]
```diff
@@ -112,9 +112,7 @@ namespace shadow_robot
       boost::shared_ptr<shadow_joints::MotorWrapper> motor_wrapper = boost::static_pointer_cast<shadow_joints::MotorWrapper>(joint_tmp->actuator_wrapper);
 
       motor_index_full = motor_wrapper->motor_id;
-      actuator_state->is_enabled_ = 1;
       actuator_state->device_id_ = motor_index_full;
-      actuator_state->halted_ = false;
```
Comment (toliver) [inline sr_robot_lib/src/sr_motor_robot_lib.cpp:117]: Well spotted.

**Comment by toliver** [comment]
It looks good to me as well. But as with the rest of the repos related to this migration to ros_control, I don't think we should merge anything to the main branch before:
- @shadowmanos finishes with the HW testing on the finger rig.
- Juan does a very thorough testing of it on a full hand.

File: sr_edc_ethercat_drivers/CMakeLists.txt (hunk 3 above)
Comment (shadowmanos) [inline sr_edc_ethercat_drivers/CMakeLists.txt]: thought it had deps on other parts of ros_ethercat_model
but I tested and it builds with ros_ethercat_hardware so I'll put that

File: sr_edc_controller_configuration/sr_edc_controller_manager.launch (hunk 2 above)
Comment (shadowmanos) [inline sr_edc_controller_configuration/sr_edc_controller_manager.launch]: The joint states will be published from a joint_state_controller with rate set elsewhere.
Mechanism statistics will be published from ros_ethercat and on second thought this isn't a bad place to set the rate so I'll uncomment and remove pr2_ prefix

File: sr_robot_lib/src/sr_muscle_robot_lib.cpp (hunk 4 above)
Comment (shadowmanos) [inline sr_robot_lib/src/sr_muscle_robot_lib.cpp]: Agreed. dynamic_casts removed on next push in both files.

File: sr_edc_controller_configuration/scripts/controllers_to_diagnostics.py (hunk 1 above)
Comment (shadowmanos) [inline sr_edc_controller_configuration/scripts/controllers_to_diagnostics.py:29]: Agreed. I'll remove import roslib from all python scripts in the folder

**Comment by shadowmanos** [comment]
Apparently ros_control and drop of pr2 dependencies will be done for ROS indigo. There is a new branch sr-ros-interface/indigo-devel with the changes of F_ros_control. Further improvements at indigo-devel may be backported to F_ros_control. F_ros_control will remain for potential use from source.
